### PR TITLE
pull/from/commit/push: retry on most failures

### PR DESCRIPTION
**Pull request by nalind**
If `PullOptions`/`BuilderOptions`/`CommitOptions`/`PushOptions`/`BuildOptions` includes a `MaxRetries` value other than 0, retry operations except for (currently) connection-refused, authentication, and no-such-repository/no-such-tag errors, at a default-but-configurable interval of 5 seconds.

Set the default for `buildah pull/from/commit/push` to 3 retries at 2 second intervals.

### CODE REVIEW

**Comment by rhatdan** [comment]
Sadly this pushed us over the travis time limit.

**Comment by rhatdan** [comment]
Do you plan to expose these options?  Should these options be handled in containers.conf?

**Comment by nalind** [comment]
I hadn't planned to expose the settings (the main thing is that the CLI opts in to it, but the API doesn't do it by default, so that OpenShift doesn't see a change in behavior, since they already have retry logic of their own), but we can if we need to.  Opinions?

**Comment by rhatdan** [comment]
I am fine without adding CLI support.

**Comment by rhatdan** [comment]
Tests are not happy.

**Comment by TomSweeneyRedHat** [comment]
Changes LGTM.  I'm guessing the changes have caused our "expect to fail" test to pause and retry a number of times causing Travis to hit the 50 minute timeout.  Prior we were completing in about 46 to 47 minutes.  The Travis timeout can't be changed unless we go with a paid plan or remove Travis entirely.

**Comment by rhatdan** [comment]
@nalind needs a rebase.

File: cmd/buildah/common.go [hunk 1]
```diff
@@ -25,6 +25,11 @@ var (
 	needToShutdownStore = false
 )
 
+const (
+	maxPullPushRetries = 3
```
Comment (rhatdan) [inline cmd/buildah/common.go]: These should probably go into containers.conf

Comment (vrothberg) [reply cmd/buildah/common.go]: The retry semantics are specific to Buildah and are not implemented in other tools. I think that it should be implemented in containers/image and be configurable in the registries.conf.

Comment (rhatdan) [reply cmd/buildah/common.go]: Ok let's merge and we can open a different PR if we decide we need this in registries.conf.

**Pull request merged**